### ISSUE TITLE
Feature: Option to ignore "Unlisted Dependencies" for ESLint/Prettier

### ISSUE THREAD

**Issue by Zamiell** [webpro/knip#315]
Hello, and thanks for the awesome tool. I have a feature request relating to meta-packages.

## The Problem

Some JavaScript/TypeScript projects use npm meta-packages in order to bundle dependencies into a single package.

For example, the `isaacscript-lint` package makes it so that instead of having to add ESLint + Prettier + a bunch of plugins + a bunch of configs as dependencies, all you have to do is to add the `isaacscript-lint` dependency. This kind of thing makes reading the dependencies in a project's "package.json" file easier, because the "real" dependencies can easily be seen and are not drowned out by 20+ different ESLint/Prettier dependencies.

Unfortunately, having an npm meta-package makes working with knip frustrating. With the default config, Knip will complain that ".eslintrc.cjs" and "prettier.config.mjs" are unused. The naïve solution is to exclude them, like this:

```jsonc
{
  "$schema": "https://unpkg.com/knip/schema.json",
  "ignore": [".eslintrc.cjs", "prettier.config.mjs"]
}
```

This solution works fine to start with, but as our projects get more complicated, we run into more problems. Say for example we want to add `eslint-plugin-sort-exports` to our linting setup. Now, Knip will complain that "eslint-plugin-sort-exports" is an unused `devDependency`. We could also include that in the `ignoreDependencies` array, but this strategy is not going to scale in the long-term.

What we really want to do is to have Knip actually parse our ESLint config. Handily, we can force Knip into knowing that our project does indeed use ESLint by doing this:

```jsonc
{
  "$schema": "https://unpkg.com/knip/schema.json",
  "eslint": true
}
```

This config gets rid of the "unused .eslintrc.cjs" error, but it then throws errors for "unlisted dependencies of eslint" (and other related errors). So we would have to ignore the unlisted dependencies with this config, which is potentially less hacky than what we started with, but not by much.

## Proposed Solution 1

It might be nice to have the ability to specify to the ESLint plugin that the dependencies really are there, like this:

```jsonc
{
  "$schema": "https://unpkg.com/knip/schema.json",
  "eslint": {
    "ignoreUnlistedDependencies": true
  }
}
```

With this flag set to true, it would parse ".eslintrc.cjs" as before, but wouldn't complain about unlisted dependencies.

Note that the Prettier plugin would also need to have this behavior, as the exact same problem applies to that.

## Proposed Solution 2

As an end-user, I should probably be able to tell Knip that I am using meta-packages in my project, like this:

```jsonc
{
  "$schema": "https://unpkg.com/knip/schema.json",
  "usingMetaPackages": true
}
```

With this option set to true, instead of just looking 1 level deep for dependencies, Knip would instead look 2 levels deep for dependencies. In other words, it would see `isaacscript-lint` as a dependency, and then look through all of the `isaacscript-lint` dependencies to see if `eslint` exists.

**Comment by webpro** [comment]
This exact phenomenon is definitely something I would love to tackle properly, it's a major concern (basically since day 1). So thanks a lot for raising this and thinking along here.

Regarding your proposed solution, I'm not targeting to ignore such issues, but rather address them as good as possible.

The `usingMetaPackages` option is an interesting option, definitely something to consider.

Another angle I'm considering is to have some kind of "transfer" mechanism that can be configured at root level like so (not sure yet about the `meta` name):

```json
{
  "eslint": { "meta": "@internal/eslint-config" },
  "jest": { "meta": "@scope/shared-configurations", "config": "config/jest.config.ts" },
  "prettier": "@internal/shared"
}
```

Knip can then "transfer" the results of a plugin to the provided meta package.

Ideally it would require no config it all, but this seems near impossible (how to recognize the meta package, etc).

Curious to what you (and others) think :)

**Comment by Zamiell** [comment]
Yes, the transfer mechanism seems like a better solution than my Proposed Solution 1.

Interestingly, the transfer mechanism would also fix an semi-related bug having to do with monorepos. Like most monorepos, I have the following setup:

```
project/project.json
project/packages/foo/package.json
```

Knip is intended to be run at the root of the monorepo, but I cannot get it working properly there, and find it easier to use Knip on individual packages at a time. However, when I run Knip from the "foo" directory, it complains that ESLint is not found, because ESLint is actually located in the root "package.json" file as a monorepo-wide dependency (which is the exact same problem in the OP, just because of monorepos instead of meta-packages).

So, if a transfer solution existed, then I could specify a config like this:

```
{
  "eslint": { "meta": "../.." },
}
```

**Comment by Zamiell** [comment]
> Ideally it would require no config it all, but this seems near impossible (how to recognize the meta package, etc).

I think it is possible in principle to detect a meta-package. For example, [here is the package.json file for isaacscript-lint](https://github.com/IsaacScript/isaacscript/blob/main/packages/isaacscript-lint/package.json). It doesn't have any "main" field or "exports" field (for code) or "types" field (for TypeScript types). So then we can presume the package "isn't doing anything" other than just being a meta-package

More formally, the algorithm would be:

1) Look through each dep package.json file.
2) If it has a "dependencies" field and does not have main/exports/types, then it is a meta-package.

Maybe there are some other fields to consider other than main/exports/types? Not sure.

**Comment by webpro** [comment]
> Knip is intended to be run at the root of the monorepo, but I cannot get it working properly there, and I find it easier to use Knip on individual packages one at a time.

You can use `knip --workspace [dir]` to focus on individual packages. The logic behind this flag has been refactored/improved around v2.32 (it includes ancestor + dependent workspaces), so you may want to (re)try that.

If that's not working well, besides what's being discussed in this thread, then please file an issue.

**Comment by webpro** [comment]
> > Ideally it would require no config it all, but this seems near impossible (how to recognize the meta package, etc).
> 
> I think it is possible in principle to detect a meta-package. For example, [here is the package.json file for isaacscript-lint](https://github.com/IsaacScript/isaacscript/blob/main/packages/isaacscript-lint/package.json). It doesn't have any "main" field or "exports" field (for code) or "types" field (for TypeScript types). So in this case, we can assume that the package "isn't doing anything" other than just being a meta-package.
> 
> More formally, the algorithm would be:
> 
> 1. Look through each dep package.json file.
> 2. If it has a "dependencies" field and does not have main/exports/types, then it is a meta-package.
> 
> Maybe there are some other fields to consider other than main/exports/types? Not sure.

Not sure if approaching it this way, taking into account the myriad of ways people using such packages, is something Knip should rely on. Granted, Knip is full of heuristics and such, but this feels a bit far fetched. There can also be more than of such packages, e.g. I've seen `@org/eslint-config` and then some package(s) with one or more (shared) configuration files (sometimes without dependencies). Not easy to link plugins to their "meta" package(s). The `main` field in `package.json` is not mandatory (default is `index.js` I think), etc.

Maybe for some plugins it could be derived from `extends` fields and such (like ESLint and Typescript configs allow), but not all tools have this.

By the way, the isaacscript monorepo is a great repository for Knip to work on this feature :)

**Comment by Zamiell** [comment]
> You can use knip --workspace [dir] to focus on individual packages. The logic behind this flag has been refactored/improved around v2.32 (it includes ancestor + dependent workspaces), so you may want to (re)try that.

The isaacscript monorepo does not use workspaces. Sorry, I know that's a bit non-idiomatic, but I think that the "correct" way for monorepos to work is to not use workspaces at all and just install all the deps at the root of the monorepo. (I have [wrote about this strategy here](https://github.com/vercel/turbo/issues/2060#issuecomment-1416135197).)

**Comment by webpro** [comment]
Knip works best from root and if it knows about the packages. Idea for the root config `knip.json`:

```json
{
  "workspaces": {
    "packages/lib-*": {},
    "packages/package-with-different-config": {}
  }
}
```

I'd recommend this, because something like `"meta": "../.."` ("external internals" to the workspace) is something I currently feel is not likely to make it into the final solution for what the original issue is about.

**Comment by Zamiell** [comment]
> Idea for the root config knip.json

Ok, that seems reasonable to me! Let me know when it's ready and I will give it a whirl. : )

**Comment by webpro** [comment]
This already works :)

**Comment by Zamiell** [comment]
Thanks webpro. To continue the discussion, I tried to run `knip` from the root of my monorepo, but I am running into some problems.

- I have a monorepo package that is an npm library. It exports helper functions in the "index.ts" file.
- These functions are otherwise unused, but obviously this is intentional because it is an npm library that other people are expected to consume.
- When running `knip` from the subpackage directory, it seems to work fine with no errors. 
- When running `knip` from the monorepo directory, it complains that all of the exports are unused.

I figured that maybe I need to explicitly tell knip about the entry point, so I tried this config, based on what you said in the previous comment:

```json
{
  "$schema": "https://unpkg.com/knip@2/schema.json",
  "workspaces": {
    "packages/foo": {
      "entry": ["./packages/foo/src/index.ts"]
    }
  }
}
```

But that didn't work. Furthermore, I changed "./packages/foo/src/index.ts" to "BOGUS_VALUE", and knip didn't even throw any errors. Is that intended? Shouldn't knip be throwing runtime errors if I put in bogus values for the purposes of failing-fast on typos? Regardless, I'm not quite sure how to make this work, any guidance would be appreciated.

**Comment by gajus** [comment]
I would really like to just be able ignore these issues, at least interim.

At the moment, there is no nice workaround.

If I am getting an error:

```ts
Unused devDependencies (7)
@graphql-codegen/add                    package.json
@graphql-codegen/typed-document-node    package.json
@graphql-codegen/typescript             package.json
@graphql-codegen/typescript-operations  package.json
@graphql-codegen/typescript-resolvers   package.json
@graphql-eslint/eslint-plugin           package.json
@roarr/cli                              package.json
Unlisted dependencies (3)
eslint-config-canonical/graphql    .eslintrc.cjs
eslint-config-canonical/zod        .eslintrc.cjs
eslint-import-resolver-typescript  .eslintrc.cjs
```

My options seem to be:

1. add `eslint-config-canonical` as a direct dependency (don't want to do that)
2. ignore `.eslintrc.cjs`

We usually do #2, but then that causes its own new issues, because now if `.eslintrc.cjs` was importing something, that suddenly gets merged as unused. So now we have to ignore these dependencies that are actually used. :-(

**Comment by webpro** [comment]
Yes, this has been awful since early days. Unfortunately I don't see a good generic solution.

The real solution is the upcoming [ESLint flat config](https://eslint.org/docs/head/use/configure/configuration-files-new) format, this will eliminate all these issues at once (already using it myself in https://github.com/webpro/knip/blob/v3/packages/eslint-config/index.js). But it will take some time before this is actually useful.

Until then, maybe it helps if Knip will support wildcards in `ignoreDependencies`. This in combination with also disabling the eslint plugin (btw this is possible today with `eslint: false`) and ignoring the ESLint config files might be somewhat bearable.

**Comment by gajus** [comment]
The problem with `{eslint: false}` is that it also starts reporting imports that were referenced in the config files as unused.

**Comment by gajus** [comment]
This one is the most annoying:

```ts
const path = require('node:path');

module.exports = {
  extends: ['@contra/eslint-config'],
  root: true,
  settings: {
    'import/resolver': {
      typescript: {
        project: path.resolve(__dirname, 'tsconfig.json'),
      },
    },
  },
};
```

Knip reports:

```bash
$ knip; knip --production --strict
Unlisted dependencies (1)
eslint-import-resolver-typescript  .eslintrc.cjs
✂️  Excellent, Knip found no issues.
```

Now if I ignore `eslint-import-resolver-typescript`:

```bash
$ at 16:24:53 ❯ knip; knip --production --strict
✂️  Excellent, Knip found no issues.
Configuration issues (1)
Unused item in ignoreDependencies: eslint-import-resolver-typescript
✂️  Excellent, Knip found no issues.
```

so cannot do neither that or that.

**Comment by webpro** [comment]
I did not resolve the actual underlying issue, but by improving a few things the workaround should help a bit here. We can now use regexes in `ignoreBinaries` and `ignoreDependencies`, and the minimal version of the workaround is explained here: https://knip.dev/guides/handling-issues/#eslint--jest

Let mew know if you encounter any issues.
